fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
         HandleMsg::SetCap { amount: _ } => Ok(HandleResponse::default()),
-        HandleMsg::SetLimits { swap_min: _, swap_max: _, swap_fee: _ } => Ok(HandleResponse::default()),
-        HandleMsg::GrantRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RevokeRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RenounceRole { role: _, address: _ } => Ok(HandleResponse::default()),
+        HandleMsg::SetLimits {
+            swap_min: _,
+            swap_max: _,
+            swap_fee: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::GrantRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RevokeRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RenounceRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
     }
 }
 
 fn try_swap<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
-    _amount: U256,
-    _destination: String,
+    env: &Env,
+    state: &State,
+    amount: Uint128,
+    destination: String,
 ) -> StdResult<HandleResponse> {
-   config(&mut deps.storage).update(|mut state| {
+    verify_not_paused(env, state)?;
+    verify_swap_amount(amount, state)?;
+
+    let swap_id = state.next_swap_id;
+    config(&mut deps.storage).update(|mut state| {
+        state.supply += amount;
         state.next_swap_id += 1;
         Ok(state)
     })?;
 
-    Ok(HandleResponse::default())
+    let log = vec![
+        log("action", "swap"),
+        log("destination", destination),
+        log("swap_id", swap_id),
+        log("amount", amount), // TOFIX(LR) how about fees?
+    ];
+
+    let r = HandleResponse {
+        messages: vec![],
+        log,
+        data: None,
+    };
+    Ok(r)
+}
+
+fn try_reverse_swap<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: &Env,
+    state: &State,
+    rid: u64,
+    to: HumanAddr,
+    from: String,
+    origin_tx_hash: Uint128,
+    amount: Uint128,
+    relay_eon: u64,
+) -> StdResult<HandleResponse> {
+    verify_tx_relay_eon(relay_eon, state)?;
+    only_relayer(env, state)?;
+
+    if amount > state.swap_fee {
+        // TOFIX(LR) when amount == fee, amount will still be consumed
+        let swap_fee = state.swap_fee;
+        let effective_amount = (amount - swap_fee)?;
+        let to_canonical = deps.api.canonical_address(&to)?;
+        let rtx = send_tokens(
+            &deps.api,
+            &env,
+            &state,
+            &to_canonical,
+            amount,
+            "reverse_swap",
+        )?;
+        config(&mut deps.storage).update(|mut state| {
+            state.supply = (state.supply - amount)?;
+            state.sealed_reverse_swap_id += 1; // TOFIX(LR) after exec should be == rid
```

Comment:
I know that we have talked about this during our design discussions, but as of now, this concept is not implemented yet in counterpart contract (ethgereum).

Reply:
@pbukva sure, I will just set it to `rid` as in the next (commented) line

Reply:
It is not enough, because:
1. By-designt here will be "holes" in reverse swaps - `reverseSwap(...)` will not be be called for all `rid`s, because swaps which were refunded on the other blockchain are never going to be tranferred to this blockchacin (= `reverseSwap(...)` will not be called for refunded `rid`s) 
2. it is not ensured (at least not in original design) that `reverseSwap(...)` will be called "nicely" in ordered sequence so that each next call to `reverseSwap(...)` will be for `rid > previous_rid` **and** that there are not other outstanding rids between `rid` passed in to current call and `previous_rid` recorded in the contract. We might implement relayer svc that way, but that is just implementation detail(specific execution strategy) of the relayer service, which is **not** enforced on contract side yet.

Reply:
understood, I'll comment it out for now then